### PR TITLE
migration: Sync unpaused condition at target pod before migration finish

### PR DESCRIPTION
**Pull request by qinqon**
### What this PR does
Before this PR:
The virt-launcher pod has a readiness gate that depends on the vm
unpaused condition to be present, this condition is set when the VM is
running, but sometimes after proper live migration readiness gate takes
some time to see the condition so there are going to be some time where
target virt-launcher unpropertly marked as not ready.

After this PR:
This change create the target pod of live migration with unpaused
condition if the VM is not paused so readiness gate is fine before live
migration finish.

It also improve a little the pod status condition patching to not fail
when there is no "conditions" fiels yet at "status.

Fixes #https://issues.redhat.com/browse/CNV-38604

### Why we need it and why it was done in this way
The following tradeoffs were made:

The following alternatives were considered:

Links to places where the discussion took place: 

### Special notes for your reviewer

TODO:
- [ ] Add test to check unpause condition creation timestamp is older than live migration finish timestamp.

### Checklist

This checklist is not enforcing, but it's a reminder of items that could be relevant to every PR.
Approvers are expected to review this list.

- [ ] Design: A [design document](https://github.com/kubevirt/community/tree/main/design-proposals) was considered and is present (link) or not required
- [ ] PR: The PR description is expressive enough and will help future contributors
- [ ] Code: [Write code that humans can understand](https://en.wikiquote.org/wiki/Martin_Fowler#code-for-humans) and [Keep it simple](https://en.wikipedia.org/wiki/KISS_principle)
- [ ] Refactor: You have [left the code cleaner than you found it (Boy Scout Rule)](https://learning.oreilly.com/library/view/97-things-every/9780596809515/ch08.html)
- [ ] Upgrade: Impact of this change on upgrade flows was considered and addressed if required
- [ ] Testing: New code requires [new unit tests](https://github.com/kubevirt/kubevirt/blob/main/docs/reviewer-guide.md#when-is-a-pr-good-enough). New features and bug fixes require at least on e2e test
- [ ] Documentation: A [user-guide update](https://github.com/kubevirt/user-guide/) was considered and is present (link) or not required. You want a user-guide update if it's a user facing feature / API change.
- [ ] Community: Announcement to [kubevirt-dev](https://groups.google.com/g/kubevirt-dev/) was considered

### Release note

```release-note
Sync unpaused condition at target pod before migration finish
```

### CODE REVIEW

**Review by davidvossel — COMMENTED** [review]
/lgtm
/approve

Yep, this makes sense to me.  We want to start the migration target pod with the same paused related conditions as the source pod in order to prevent a brief lag in network availability. 

This does introduce a brief bit of risk in the event that someone pauses/unpauses a VMI right as the migration is taking place. In that case, the target pod might start with an inaccurate condition, but that will immediately converge to the correct condition after the migration completes. I don't think it's important to optimize for this possibility. Starting the target pod with the source pods condition closes the primary issue without making this any worse.

File: pkg/virt-controller/watch/migration.go [hunk 1]
```diff
@@ -727,6 +727,14 @@ func (c *MigrationController) createTargetPod(migration *virtv1.VirtualMachineIn
...
 		return err
 	}
+
+	// If VM was not paused patch the pod with unpaused condition
+	// to allow readiness gate to catch up before live migration is finished
+	if err := syncPausedConditionToPod(c.clientset, vmi, pod); err != nil {
+		c.recorder.Eventf(vmi, k8sv1.EventTypeWarning, FailedCreatePodReason, "Error syncing paused condition at migration at vmi target pod: %v", err)
+		return fmt.Errorf("failed to sync paused condition at vmi migration target pod: %v", err)
+	}
+
```
Comment (davidvossel) [inline pkg/virt-controller/watch/migration.go:737]: this needs to be moved outside of the createTargetPod function. The reason is, the create could be successful, but this patch could fail. The patch will never be eventually consistent in that case because this function isn't called if the pod already exists.

**Review by davidvossel — COMMENTED** [review]
/lgtm cancel

sorry, caught something minor after the first review

**Review by EdDev — COMMENTED** [review]
I think this change merits another reviewer from the sig-compute.

> We want to start the migration target pod with the same paused related conditions as the source pod

The implementation seems to recalculate it, not to read it from the source pod. But maybe I am missing some parts.

Generally speaking, I find this status update wrong. The migration controller has no business to reflect something like this prematurely. While the migration controller creates the pod, the vmi controller reconciles it. Feels like a workaround to a real problem.

Nevertheless, I am not familiar with all the details and do not have an alternative to suggest.

**Comment by qinqon** [comment]
> I think this change merits another reviewer from the sig-compute.
> 
> > We want to start the migration target pod with the same paused related conditions as the source pod
> 
> The implementation seems to recalculate it, not to read it from the source pod. But maybe I am missing some parts.
> 

It recalculate from VMI status not from source pod, I will reword this.

File: pkg/virt-controller/watch/vmi.go [hunk 2]
```diff
@@ -816,8 +816,12 @@ func preparePodPatch(oldPod, newPod *k8sv1.Pod) ([]byte, error) {
 			return nil, err
 		}
 
-		patchOps = append(patchOps, fmt.Sprintf(`{ "op": "test", "path": "/status/conditions", "value": %s }`, string(oldConditions)))
-		patchOps = append(patchOps, fmt.Sprintf(`{ "op": "replace", "path": "/status/conditions", "value": %s }`, string(newConditions)))
+		if string(oldConditions) == "null" {
```
[the remark this answers is not in the document]
Comment (qinqon) [reply pkg/virt-controller/watch/vmi.go:819]: There is also the issue of "test" failing if "condition" field do not exist yet even if "test" has `"value": null`, so in those case, conditions shoul de direclty added, this replicates the fix at https://github.com/kubevirt/kubevirt/blob/main/pkg/virt-controller/watch/vmi.go#L842.

File: pkg/virt-controller/watch/migration.go [hunk 3]
```diff
@@ -727,6 +727,14 @@ func (c *MigrationController) createTargetPod(migration *virtv1.VirtualMachineIn
 		c.podExpectations.CreationObserved(key)
 		return err
 	}
+
+	// If VM was not paused patch the pod with unpaused condition
+	// to allow readiness gate to catch up before live migration is finished
+	if err := syncPausedConditionToPod(c.clientset, vmi, pod); err != nil {
+		c.recorder.Eventf(vmi, k8sv1.EventTypeWarning, FailedCreatePodReason, "Error syncing paused condition at migration at vmi target pod: %v", err)
```
[the remark this answers is not in the document]
Comment (qinqon) [reply pkg/virt-controller/watch/migration.go:734]: Right, I will remove it.

File: pkg/virt-controller/watch/vmi.go [hunk 4]
```diff
@@ -994,7 +998,7 @@ func (c *VMIController) syncReadyConditionFromPod(vmi *virtv1.VirtualMachineInst
 	}
 }
 
-func (c *VMIController) syncPausedConditionToPod(vmi *virtv1.VirtualMachineInstance, pod *k8sv1.Pod) error {
+func syncPausedConditionToPod(clientset kubecli.KubevirtClient, vmi *virtv1.VirtualMachineInstance, pod *k8sv1.Pod) error {
```
[the remark this answers is not in the document]
Comment (qinqon) [reply pkg/virt-controller/watch/vmi.go:1001]: I will move it it out.

**Comment by xpivarc** [comment]
> I think this change merits another reviewer from the sig-compute.
> 
> > We want to start the migration target pod with the same paused related conditions as the source pod
> 
> The implementation seems to recalculate it, not to read it from the source pod. But maybe I am missing some parts.
> 
> Generally speaking, I find this status update wrong. The migration controller has no business to reflect something like this prematurely. While the migration controller creates the pod, the vmi controller reconciles it. Feels like a workaround to a real problem.
> 
> Nevertheless, I am not familiar with all the details and do not have an alternative to suggest.

+1

@EdDev has great remarks in his review. Please, also make sure you cover this change by unit tests.

File: pkg/virt-controller/watch/migration.go (hunk 1 above)
Comment (qinqon) [reply pkg/virt-controller/watch/migration.go:737]: I will move it to `updateStatus`

File: pkg/virt-controller/watch/vmi.go (hunk 2 above)
[the remark this answers is not in the document]
Comment (qinqon) [reply pkg/virt-controller/watch/vmi.go]: @EdDev, after checking again this is not necessary, I will remove this.

**Comment by qinqon** [comment]
@davidvossel @EdDev @xpivarc can you take another look ?

**Comment by vladikr** [comment]
> /lgtm /approve
> 
> Yep, this makes sense to me. We want to start the migration target pod with the same paused related conditions as the source pod in order to prevent a brief lag in network availability.
> 
> This does introduce a brief bit of risk in the event that someone pauses/unpauses a VMI right as the migration is taking place. In that case, the target pod might start with an inaccurate condition, but that will immediately converge to the correct condition after the migration completes. I don't think it's important to optimize for this possibility. Starting the target pod with the source pods condition closes the primary issue without making this any worse.

It makes sense to me as well. 
We had a discussion about this on [slack](https://kubernetes.slack.com/archives/C0163DT0R8X/p1708602920325019) some time ago.
@rmohr had some concerns.

**Comment by davidvossel** [comment]
> Generally speaking, I find this status update wrong. The migration controller has no business to reflect something like this prematurely. While the migration controller creates the pod, the vmi controller reconciles it. Feels like a workaround to a real problem.

It's an approximation to close the gap of a real problem.

The problem is that the presence of this condition on a Pod influences the Pod's readiness state. This readiness is what Kubernetes Services use to decide whether traffic flows to an endpoint or not.

So, without the immediate presence of this condition on the target pod, when live migration completes there's a race condition. If the vmi controller does not update this condition quickly enough after migration completes, then the service will see the target pod as not ready, this results in brief drop in traffic. When the VM in question is a Kubernetes node (cluster-api-provider-kubevirt use case), this means there's a network disruption that occurs during live migration.  We this breaks our conformance tests and is a real issue.

By calculating and setting this condition when the target pod is created, we are giving an initial "approximation" of what we think this condition will be when the target becomes the source pod. When the migration completes, the vmi controller will converge this approximation with whatever the reality is. By approximating this at creation time, we're able to close the readiness gap (like > 99.99% of the time) with the caveat being that the approximation would be inaccurate if someone paused/unpaused the vmi immediately after the approximation took place. Then the worst case is that the target pod's condition is immediately re converged after the migration completes.

The vmi controller does not reconcile pending migration target pods, which is why the best approximation for this condition is to do this calculation in the migration controller.

**Comment by davidvossel** [comment]
/hold

I want to better understand the implications of having both the target and source pod "ready" at the same time before this merges

**Comment by qinqon** [comment]
> /hold
> 
> I want to better understand the implications of having both the target and source pod "ready" at the same time before this merges



> /hold
> 
> I want to better understand the implications of having both the target and source pod "ready" at the same time before this merges

To add more context, without this PR endpointslices are already both with ready and serving true in the middle of the migration 

```
Tue Feb 27 05:25:53 PM CET 2024                                                 
[                                                                               
  {                                                                             
    "addresses": [                                                              
      "10.244.2.10"                                                             
    ],                                                                          
    "conditions": {                                                             
      "ready": true,                                                            
      "serving": true,                                                          
      "terminating": false                                                      
    },                                                                          
    "nodeName": "ovn-worker",                                                   
    "targetRef": {                                                              
      "kind": "Pod",                                                            
      "name": "virt-launcher-worker1-5l746",                                    
      "namespace": "default",                                                   
      "uid": "116c5fb1-65bc-4551-84a2-e2d4de862d3d"                             
    }                                                                           
  },                                                                            
  {                                                                             
    "addresses": [                                                              
      "10.244.1.7"                                                              
    ],                                                                          
    "conditions": {                                                             
      "ready": true,                                                            
      "serving": true,                                                          
      "terminating": false                                                      
    },                                                                          
    "nodeName": "ovn-worker2",                                                  
    "targetRef": {                                                              
      "kind": "Pod",                                                            
      "name": "virt-launcher-worker1-v4pj8",                                    
      "namespace": "default",                                                   
      "uid": "a8f24743-747e-4c50-9117-2cd265957500"                             
    }                                                                           
  }                                                                             
]                                                  
```

**Comment by qinqon** [comment]
At vanilla kubevirt with normal masquerade VM both virt-launchers are ready at the same time too 
```
10:23:20.377410729                                                              
{                                                                               
  "lastProbeTime": null,                                                        
  "lastTransitionTime": "2024-02-29T09:22:51Z",                                 
  "message": "corresponding condition of pod readiness gate \"kubevirt.io/virtual-machine-unpaused\" does not exist.",
  "reason": "ReadinessGatesNotReady",                                           
  "status": "False",                                                            
  "type": "Ready"                                                               
}                                                                               
{                                                                               
  "lastProbeTime": null,                                                        
  "lastTransitionTime": "2024-02-29T09:16:53Z",                                 
  "status": "True",                                                             
  "type": "Ready"                                                               
}                                                                               
10:23:20.659527191                                                              
{                                                                               
  "lastProbeTime": null,                                                        
  "lastTransitionTime": "2024-02-29T09:23:20Z",                                 
  "status": "True",                                                             
  "type": "Ready"                                                               
}                                                                               
{                                                                               
  "lastProbeTime": null,                                                        
  "lastTransitionTime": "2024-02-29T09:16:53Z",                                 
  "status": "True",                                                             
  "type": "Ready"                                                               
}                                                                               
10:23:20.922261579                                                              
{                                                                               
  "lastProbeTime": null,                                                        
  "lastTransitionTime": "2024-02-29T09:23:20Z",                                 
  "status": "True",                                                             
  "type": "Ready"                                                               
}                                                                               
{                                                                               
  "lastProbeTime": null,                                                        
  "lastTransitionTime": "2024-02-29T09:23:20Z",                                 
  "message": "containers with unready status: [guest-console-log]",             
  "reason": "ContainersNotReady",                                               
  "status": "False",                                                            
  "type": "Ready"                                                               
}                                 
```

**Comment by qinqon** [comment]
With this PR this is how virt-launcher source and target pods ready condition looks during switch over (it has a 0.2 seconds refresh)

```
3:24:44.251882815                                                              
{                                                                               
  "lastProbeTime": null,                                                        
  "lastTransitionTime": "2024-03-01T12:23:56Z",                                 
  "message": "corresponding condition of pod readiness gate \"kubevirt.io/virtual-machine-unpaused\" does not exist.",
  "reason": "ReadinessGatesNotReady",                                           
  "status": "False",                                                            
  "type": "Ready"                                                               
}                                                                               
{                                                                               
  "lastProbeTime": null,                                                        
  "lastTransitionTime": "2024-03-01T12:23:10Z",                                 
  "status": "True",                                                             
  "type": "Ready"                                                               
}                                                                               
13:24:44.529730973                                                              
{                                                                               
  "lastProbeTime": null,                                                        
  "lastTransitionTime": "2024-03-01T12:24:44Z",                                 
  "status": "True",                                                             
  "type": "Ready"                                                               
}                                                                               
{                                                                               
  "lastProbeTime": null,                                                        
  "lastTransitionTime": "2024-03-01T12:23:10Z",                                 
  "status": "True",                                                             
  "type": "Ready"                                                               
}                                                                               
13:24:44.818707344                                                              
{                                                                               
  "lastProbeTime": null,                                                        
  "lastTransitionTime": "2024-03-01T12:24:44Z",                                 
  "status": "True",                                                             
  "type": "Ready"                                                               
}                                                                               
{                                                                               
  "lastProbeTime": null,                                                        
  "lastTransitionTime": "2024-03-01T12:23:10Z",                                 
  "status": "True",                                                             
  "type": "Ready"                                                               
}
13:24:45.089195243                                                              
{                                                                               
  "lastProbeTime": null,                                                        
  "lastTransitionTime": "2024-03-01T12:24:44Z",                                 
  "status": "True",                                                             
  "type": "Ready"                                                               
}                                                                               
{                                                                               
  "lastProbeTime": null,                                                        
  "lastTransitionTime": "2024-03-01T12:24:45Z",                                 
  "message": "containers with unready status: [volumecontainerdisk guest-console-log]",
  "reason": "ContainersNotReady",                                               
  "status": "False",                                                            
  "type": "Ready"                                                               
}                                             
```

**Comment by EdDev** [comment]
> The vmi controller does not reconcile pending migration target pods, which is why the best approximation for this condition is to do this calculation in the migration controller.

This would be a better fix IMO. If the VMI controller can reconcile also the target pod and add the condition based on some information that is relevant.

> To add more context, without this PR and in normal situations, endpointslices are already both with ready and serving true in the middle of the migration

I would expect this to be bad as well, but the current change making it a bit worse.
I admit not to understand how things work so well, but if the new pod declares itself as able to server the service, new redirections to it will fail.

I guess the ideal is to make the source and target switch as fast as possible with.
As the VMI controller is the one that "sees" both pods, I think it will be best for it to explicitly implement the policy (either keep them both with the condition or without)
I would prefer the migration controller not to know or care about this in any way.

I am also unsure you adds the condition itself (is it the VMI controller?), but I am pretty sure that the information of connectivity availability is not guaranteed when posting the pod manifest.
But I do understand the attempt to reverse the condition/reediness default, I just think it is a hack which we better avoid unless we exhausted all other options.

File: pkg/virt-controller/watch/util/util.go [hunk 5]
```diff
@@ -118,3 +123,80 @@ func CreateDataVolumeManifest(clientset kubecli.KubevirtClient, dataVolumeTempla
 
 	return newDataVolume, nil
 }
+
+func SyncPausedConditionToPod(clientset kubecli.KubevirtClient, vmi *virtv1.VirtualMachineInstance, pod *corev1.Pod) error {
```
Comment (EdDev) [inline pkg/virt-controller/watch/util/util.go:127]: I would prefer this will be placed in a better specific location over under a generic nameless `util`.

Can this fit in "kubevirt.io/kubevirt/pkg/controller" ?

Comment (qinqon) [reply pkg/virt-controller/watch/util/util.go:127]: Right, also is only use by `pkg/virt-controller/watch/`   so it do not make sense to make it public, what do you think about move this to `pkg/virt-controller/watch/sync.go` and make it private again ?

**Comment by qinqon** [comment]
> > The vmi controller does not reconcile pending migration target pods, which is why the best approximation for this condition is to do this calculation in the migration controller.
> 
> This would be a better fix IMO. If the VMI controller can reconcile also the target pod and add the condition based on some information that is relevant.
> 
> > To add more context, without this PR and in normal situations, endpointslices are already both with ready and serving true in the middle of the migration
> 
> I would expect this to be bad as well, but the current change making it a bit worse. I admit not to understand how things work so well, but if the new pod declares itself as able to server the service, new redirections to it will fail.

@EdDev I have tune the PR a little, now it set the unpause condition only when the Migration state that target domain is ready to receive traffic, after taking some numbers both pods are ready < 1s.

> 
> I guess the ideal is to make the source and target switch as fast as possible with. As the VMI controller is the one that "sees" both pods, I think it will be best for it to explicitly implement the policy (either keep them both with the condition or without) I would prefer the migration controller not to know or care about this in any way.
> 
> I am also unsure you adds the condition itself (is it the VMI controller?), but I am pretty sure that the information of connectivity availability is not guaranteed when posting the pod manifest. But I do understand the attempt to reverse the condition/reediness default, I just think it is a hack which we better avoid unless we exhausted all other options.
